Have clients upload report shares to helpers (#174)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-gpew-priv-ppm.md

[[OPEN ISSUE: This shows two helpers, but the document only allows one for now.
https://github.com/abetterinternet/ppm-specification/issues/117]]

~~~
[NOTE: This is a potential alternative to {{pa-topology}} that just shows how
parties are connected, i.e., who make requests to whom. Note that there may be
mulitple instances of the "Helper" role.]
+------------+              +------------+
|            | upload       |            |
|   Client   +-------------->   Leader   |
|            |              |            |
+-----+------+              +--+---^-----+
      | upload       aggregate |   |
      |                        |   |
      |   +--------------------+   |
      |   |                        |
      |   |                        | collect
+-----V---V--+              +------+-----+
|            |              |            |
|   Helper   |              | Collector  |
|            |              |            |
+------------+              +------------+
~~~

> {: #pa-topology-alt title "System Architecture"}

Reviewer: Agree with Tim here -- I'm not seeing the value of keeping both.

refer to it in protocol messages.

During the duration of the measurement, each client records its own value(s),
packages them up into a sequence of report shares, and sends one share to each
aggregator. Each share is encrypted to the intended recipient so that even if
report shares are handled by an intermediary, that interdmediary cannot recover
the client's measurement.

The leader orchestrates the process of verifying them (see
{{validating-inputs}}) and assembling them into a final aggregate for the
collector. Depending on the VDAF, it may be possible to incrementally process
each report as it comes in, or it may be necessary to wait until the entire

> batch of reports is received.

Author: No that's not possible. Yeah, it's the aggregation parameter that matters, but here (in the overview) we have so far avoided talking about the aggregation parameter.

Note that this change isn't new to this PR.

> batch of reports is received.

Reviewer 2: OK -- there's enough to chew on in this PR that we don't have to wordsmith this right now.

* `aggregator_endpoints`: A list of URLs relative to which an aggregator's API
  endpoints can be found. Each endpoint's list MUST be in the same order. The

>   leader's endpoint MUST be the first in the list.

Reviewer: Is this requirement necessary anymore? It seems like the client doesn't need to care which aggregator is the leader if both are treated as separate aggregators to store shares.

Reviewer: Oh, yeah, that was totally not clear to me! How is this expressed in terms of VDAF?

Author: It's not explicit in the syntax, but for schemes that needs to differentiate aggregator roles, they can do so via the verification parameter. For example, see https://cjpatton.github.io/vdaf/draft-patton-cfrg-vdaf.html#name-setup-2

## Uploading Reports

Clients periodically upload report shares to the aggregators. This involves two
HTTP requests, one to retrieve the aggregator's HPKE configuration and
another to upload its report share.

### Key Configuration Request {#key-config}

> Before the client can upload a report share to an aggregator, it must first
> obtain the aggregator's public key. This is retrieved by sending a request to

Reviewer: ```suggestion
Before the client can upload a report share to an aggregator, it first fetches
the aggregator's public key. This is done by sending a request to
```

### Upload Request

Report shares are usually uploaded by clients directly, but MAY instead be
uploaded by an intermediary that ingests and buffers reports. A report share is

> uploaded using an HTTP POST to `[aggregator]/upload`, where `[aggregator]` is

Reviewer: This seems to preclude sending both reports to the same endpoint, even though they're encrypted under separate aggregator public keys. We could address this by allowing the task configuration to specify the upload endpoint(s) _separately_ from the aggregation endpoints. Would that work?

Reviewer: Well, consider the case where there is an ingestor service. Certainly, that ingestor service can't be the _client_ in PPM, since that would require it see inputs and then encrypt them under the aggregator public keys. Rather, the ingestor sees encrypted things that it then relays onward to the aggregator endpoints. What we're specifying here is the client behavior, and I think we need to make sure that we the client can send both its encrypted shares to a single ingestor if needed. Right now that seems prohibited?

Author: Ok I see. Then I think this section would need to be clear that it specifies both the behavior of the client generating report shares and the entity (either the client or someone else) that uploads shares to the aggregators. Would you prefer these be different sections perhaps?

### Upload Request

Report shares are usually uploaded by clients dirctly, but MAY instead be
uploaded by an intermediary that ingests and buffers reports on behalf of the

> collector. A report share is uploaded using an HTTP POST to

Author: IMO this is out-of-scope. The way we should try to spell this PR is to allow either clients to upload report shares directly or allow report shares to be uploaded by an intermediary. How the report shares get to the intermediary isn't relevant for us.

> collector. A report share is uploaded using an HTTP POST to

Author: Good point, that's just one possible case. I dropped "on behalf of the collector".

> collector. A report share is uploaded using an HTTP POST to

Reviewer 2: This PR, sure, but I'm not the only person left wondering whether or how intermediates can batch up reports: ekr asked about this on the PPM list. Anyway, I filed #188 so we don't lose track of the idea.

where `input_share` is the aggregator's input share and `nonce` and `extensions`
are the corresponding fields of `Report`.

An aggregator responds to well-formed requests to `[aggregator]/upload` with
status 200 and an empty body. Malformed requests are handled as described in
{{errors}}. Clients SHOULD NOT upload the same measurement value in more than
one report if the aggregator responds with status 200 and an empty body.

The aggregator responds to requests with out-of-date `HpkeConfig.id` values,
indicated by `EncryptedInputShare.config_id`, with status 400 and an error of
type 'outdatedConfig'. Clients SHOULD invalidate any cached aggregator
`HpkeConfig` and retry with a freshly generated Report. If this retried report
does not succeed, clients MUST abort and discontinue retrying.

To prevent reports from being misused, the aggreagtors filter out replayed or
otherwise stale reports during the aggregation protocol using the procedure
described in {{anti-replay}}. In addition, the aggregator that receives a stale
report SHOULD abort the upload protocol and alert the client with error
"staleReport".

> ### Upload Extensions {#upload-extensions}

Reviewer 2: The reason we had upload extensions in the first place was so that they could be tunnelled through the leader into the helper. Now that we do direct uploads, can we remove these? The obvious remaining use case is tunnelling extensions through a client proxy/intermediary/batching client, but if deployments use OHTTP as we recommend, then PPM shouldn't need to specify anything as things like HTTP request headers will get securely relayed.

Author: #89 will likely make use of upload extensions, hence we'll need to keep them around.

Reviewer: We may also need general extensibility outside of new VDAFs in the future -- a lesson learned from TLS.

Reviewer 2: Well, I don't want to open the can of worms of client attestation in the margins of this PR, but I'm happy to accept TLS as a precedent.

### Aggregate Request

The AggregateReq request is used by the leader to coordinate aggregation of a
set of reports with the helpers. These reports incident to a particular

> AggregateReq MUST all be associated with the same PPM task and batch.

Reviewer 2: Now that the helper is responsible for its own nonce, report and aggregate accumulator storage, do we need aggregate requests to not span batch intervals anymore? I'm not sure but I think the reason we had this constraint on aggregate requests was so that the state blob could be "sharded" by batch interval and allow parallelism in the helper. Now, though, we might expect the helper to have its own storage where it keeps track of aggregate accumulators for several different batch intervals.

Author: Yeah, it may not be strictly necessary. However I think we should keep this as-is for now in order to keep this PR minimal. Plus, it may be helpful for something we're not seeing right now.

Reviewer: Agreed. Let's note this and revisit in a followup change. (Good catch, Tim!)

  TaskID task_id;
  opaque agg_param<0..2^16-1>;  // VDAF aggregation parameter
  opaque helper_state<0..2^16>; // helper's opaque state
  AggregateSub seq<1..2^24-1>;
} AggregateReq;
~~~

[TODO: Split into AggregateReq and AggregateInitReq so that only the latter
carries the aggregation parameter.]

The structure contains the PPM task, an opaque, VDAF-specific aggregation
parameter, an opaque *helper state* string, and a sequence of *sub-requests*,
each corresponding to a unique client report. Sub-requests are structured as
follows:

~~~
struct {
  Nonce nonce;               // The report nonce
  opaque message<0..2^16-1>; // VDAF message
} AggregateSub;
~~~

The `nonce` field has the same value as those in the leader's report share

> uploaded by the client. The remainder of the structure is dedicated to

Reviewer: Hmm, now that the nonce<>client uniqueness is no longer enforced by a single party, what happens if, say, two benign clients happen to generate colliding nonce values, and each aggregator picks a different one while filtering for anti-replay? We could maybe increase the width of the nonce random value to hedge against this collision, but it's not clear to me that's sufficient.

Author: Oh interesting! To be clear, before #169 we still had both aggregators enforce anti-replay. The difference is that we no longer of single party pick the set of reports that are processed. Nice catch.

This basically amounts to an attack on correctness. What will happen for prio3 or poplar1 is that the aggregators will derive different verification randomness, which will cause them to deem the report invalid and reject it (with high probability).

However, where this edge case *would* matter is if a VDAF were used that didn't have any verifiability. (We should probably not get in the habbit of calling such a scheme a VDAF: see https://github.com/cjpatton/vdaf/issues/20.) This was already requested in #45, so it's within the realm of possibility.

Note that a collision fairly likely, since the nonce only contains 64 random bits. (The remaining bits are a timestamp.) We may consider bumping this to 128 just to be on the safe side.

How about addressing this by leaving an OPEN ISSUE in the text? It would reference #45 and we would leave a note there to make sure we keep track of this.

Reviewer: That's a fine resolution, yep. Let's also note that this type of collision will cause verification failures with overwhelming probability?

Author: Will do.

carries the aggregation parameter.]

The structure contains the PPM task, an opaque, VDAF-specific aggregation
parameter, an opaque *helper state* string, and a sequence of *sub-requests*,
each corresponding to a unique client report. Sub-requests are structured as
follows:

~~~
struct {
  Nonce nonce;               // The report nonce
  opaque message<0..2^16-1>; // VDAF message
} AggregateSub;
~~~

The `nonce` field has the same value as those in the leader's report share
uploaded by the client. The remainder of the structure is dedicated to
VDAF-specific request parameters.

In order to provide replay protection, the leader preprocesses the set of
reports incident to the AggregateReq as described in {{anti-replay}}. Any
reports filtered out by this procedure MUST be ignored.

The helper handles well-formed requests as follows. (As usual, malformed
requests are handled as described in {{errors}}.) It first looks for PPM
parameters corresponding to `AggregateReq.task_id`. It then preprocesses the
reports incident to the AggregateReq as described in {{anti-replay}}. For any
report filtered out by this procedure, the helper MUST respond with an error

> message rather than its next VDAF message.

Reviewer 2: ```suggestion
sub-response rather than its next VDAF message.
```
To make it clear that filtering out an individual report does not fail the overall aggregate transaction.

### Aggregate Share Request {#aggregate-share-request}

Once it has verified at least as many reports as required for the PPM task, the
leader may issue an "aggregate-share request" to each helper. The helper

> responds to this request by extracting its aggregate share from its state and

Reviewer 2: nit: "extracting its aggregate share from its state" implies the helper will have an aggregate share ready to go in its state, which won't be true for `poplar1`. We could just say "The helper responds to this request with its aggregate share, encrypted under the collector's HPKE public key."

Author: Actually at this point it will have computed an aggregate share. However I took your suggestion because your text is a lot clearer.

~~~
struct {
  HpkeCiphertext encrypted_output_share;
} AggregateShareResp;
~~~

The leader uses the helper's AggregateShareResp to respond to the collector's
collect request (see {{pa-collect}}).

After issuing an aggregate-share request for a given batch interval, it is an

> error for the leader to issue any more aggreagte requests for reports in the

Reviewer 2: ```suggestion
error for the leader to issue any more aggregate requests for reports in the
```

batch interval. These reports will be rejected by helpers as described in

> {{aggregate-request}}.

Reviewer 2: It's an error for the leader to issue more than `max_batch_lifetime` requests for a given batch interval, right? As written this states the leader can issue exactly one request.

Author: Duplicate of https://github.com/abetterinternet/ppm-specification/pull/174#discussion_r791074329?

pertaining to reports that were previously aggregated for a given task. Before
processing a report, an aggregator first checks if the report pertains to a
batch that has already been "collected". This is the case if the aggregator is
the leader and has completed a collect request for a batch containing the report
(see {{pa-collect}}) or if the aggregator is a helper and has completed an
aggregate-share request for a batch containing the report (see
{{aggregate-share-request}}). It also checks if the report has been
"aggregated". This is the case if a report with the same nonce was previously
used in an aggregate request.

* If the report has not yet been aggregated but is contained by a batch that has
  been collected, then the aggregator MUST ignore it. This prevents additional
  reports from being aggregated after their batch has already been collected.

* If the report has been aggregated but has not been collected, then the
  aggregator MUST ignore it. This prevents a report from being used more than
  once in a batch.

In particular, an aggregator only process a report if either it has been
aggregated and collected or it has never been aggregated and is not contained by

> a batch that has been collected.

Reviewer: This new requirement is pretty subtle to me. Can we drop an open issue here to further investigate if this is indeed the correct thing to do?

Author: I agree it's subtle. I'd be good with that outcome, although it was already discussed here in a different context: https://github.com/abetterinternet/ppm-specification/issues/183#issuecomment-1016978627.

Reviewer: Let's at least leave an OPEN ISSUE referencing #183, then, since that's still open. (I will review #183 more carefully soon.)

detect and filter out replayed reports.

To detect replay attacks, each aggregator keeps track of the set of nonces
pertaining to reports that were previously aggregated for a given task. Before
processing a report, an aggregator first checks if the report pertains to a
batch that has already been "collected". This is the case if the aggregator is
the leader and has completed a collect request for a batch containing the report
(see {{pa-collect}}) or if the aggregator is a helper and has completed an
aggregate-share request for a batch containing the report (see
{{aggregate-share-request}}). It also checks if the report has been
"aggregated". This is the case if a report with the same nonce was previously
used in an aggregate request.

* If the report has not yet been aggregated but is contained by a batch that has

>   been collected, then the aggregator MUST ignore it. This prevents additional
>   reports from being aggregated after its batch has already been collected.

Reviewer 2: ```suggestion
  been collected, then the aggregator MUST ignore it. This prevents additional
  reports from being aggregated after their batch has already been collected.
```

* If the report has been aggregated but has not been collected, then the
  aggregator MUST ignore it. This prevents a report from being used more than

>   once in a batch.

Reviewer 2: I think but am not certain that this needs to reference the task's `max_batch_lifetime` rather than allowing exactly once, because successive collector queries will change the aggregation parameter (e.g., different prefixes in `poplar1`).

Author: I don't think want to change the batch after it has been collected with *any* aggregation parameter. Imagine, for example, that the collector requests the aggregate for candidate prefixes ["ab"]. Then add another measurement and re-run with candidate prefixes ["abc"]. Then I know immediately whether the additional measurement was prefixed by "abc".

Reviewer 2: Then I think there's something I must not be understanding about how `max_batch_lifetime` works, or how PPM is meant to solve t-heavy-hitters, which I thought involved the collector sending a series of collect requests with longer candidate prefixes each time. But that doesn't need to be solved in this PR.

Author: You understand correctly how PPM is supposed to solve heavy hitters. All `max_batch_lifetime` is supposed to say is that you can't make any more collect requests than `max_batch_lifetime`. Otherwise it doesn't make any restrictions on the aggregation parameter. In particular, Poplar can be implemented as you describe.

All that this line is saying is that you can't change the set of reports in a batch after you've collected it. Otherwise, this could lead to a privacy violation, regardless of the VDAF in use.
